design-proposals/klip-11-DQL.md

```diff
@@ -1 +1,745 @@
+* we might not need to implement the new `AS OF <timestamp>` clause in the first iteration
+* we might not need to support terminating queries over streams right away
+
+We limit this KLIP to not introduce too much syntactic sugar to express queries.
+This may make some queries unnecessary "complex/clumsy" but the goal is to illustrate to underlying principles of the language.
+We can add syntactic sugar after we agree on the basic design of the language.
+
+
+
+## Value/Return
+
+The "Dual Query Language" should be easier to understand and easier to use.
+It will allow to support nested queries seamlessly and unifies query semantics (persistent vs. transient queries).
+Furthermore, the language is designed to be "future proof", i.e., it considers potential future features and thus allows
+to extend the capabilities of KSQL without the need to change the language in a breaking way.
+
+
+
+## Public APIS
+
+We propose the following language changes:
+
+* introduce `TABLE(<stream>)` operator: takes a STREAM as input, and changes its semantic interpretation from facts to updates; it returns a TABLE
+  * note: the returned TABLE is not an RDBMS style table, but rather a MV
+  * note: it's not specified if the returned TABLE is materialized into a state store or represented as a changelog
+    (it's up the the KSQL optimizer or KS runtime to make a materialization decision)
+  * the `TABLE()` operator consumes it's input STREAM "from beginning"
+  * we could extend `TABLE()` to also accept a TABLE as input if it makes the language smoother;
+    it's obvious that `TABLE(<table>)` would be idempotent, i.e., `TABLE(t) = t`
+* introduce `STREAM(<table>)` operator: takes a TABLE as input, and returns a STREAM that contains a record for each update to the TABLE
+  * note: the returned STREAM is a fact stream, and would be processed with corresponding STREAM semantics (it's semantically not a changelog stream)
+  * note: the input TABLE can either be an RDBMS style table of a MV
+  * by default, the output STREAM contains one record for each input table record (full table scan on the table generation when the query was issues)
+    plus a record for each input table update
+  * we could extend `STREAM()` to also accept a STREAM as input if it makes the language smoother;
+  it's obvious that `STREAM(<stream>)` would be idempotent, i.e., `STREAM(s) = s`
+* Introduce an `AS OF <timestamp>` clause for RDBMS style queries against tables
+* Introduce `CREATE MATERIALIZED VIEW` statement
+* Deprecate `CREATE TABLE` statement (in favor of the new `CREATE MATERIALIZED VIEW` statement) and eventually remove it
+* Deprecate `EMIT CHANGES` clause
+
+Semantic changes:
+
+* An aggregation query over a stream returns a stream (and not a table)
+
+Naming:
+
+* Drop the terms `pull query` and `push query`
+* Use the terms `table query` and `stream query`
+  (this will also allow to use the phrase "query a stream" and "query a table" in a straight forward way)
```

Comment:
I would phrase it differently: If you have a table, you can either query its state, or you subscribe to its changelog stream. The former is "querying", while the later is a "subscribing". However, you don't subscribe to the _TABLE_, but to its _changelog_, ie, you first create a stream, via `STREAM(<table>)` and then subscribe to it.

For stream, you can only subscribe to them.

Does this make sense?

Reply:
I can be convinced - I think what you're saying makes sense that you don't "subscribe to a table" but rather to the changelog of it, I'm just considering a "push query" (indicated by something like `EMIT ...`) would do the implicit cast for you. That way you don't run into the concern I raised here: https://github.com/confluentinc/ksql/pull/3799#discussion_r349279356

Reply:
Not sure if I understand you last point. Can you elaborate? What does `EMIT` allow you to do, that you cannot do via `STREAM(table)` ?

Reply:
I brought up a similar aspect during the EMIT CHANGES discussion to just keep one mode per entity (stream/table). I think the reasoning to push back was - we already support subscribing to a table. An extra conversion to `STREAM(table)` will simplify the language . +1

Reply:
> However, you don't subscribe to the TABLE, but to its changelog, ie, you first create a stream, via STREAM(<table>) and then subscribe to it.

[Temporal tables](https://docs.microsoft.com/en-us/sql/relational-databases/tables/temporal-tables?view=sql-server-ver15), a standard SQL feature, blur the lines between a 'current value only' TABLE (which is what @mjsax refers to here) and what the history or changelog of that table is.

I think we should re-evaluate KLIP-11 in a way that aligns more with the existing, non-streaming SQL standard in this context.

Reply:
Not sure why temporal tables would be a problem? You just just refer to an older version of the table via `AS OF <timestamp>` clause that is mentioned in the KLIP. If you want to get the table starting from yesterday and stream it's changelog since then, you do `STREAM(table AS OF <timestamp>)` or the full syntax:
```sql
SELECT * FROM STREAM(SELECT * FROM table AS OF <timestamp>)
```

Btw: I had a couple of in person discussion and (_I think_) I gained a better understanding what "push and pull" is supposed to mean (what is quite unclear from KLIP-8 or from the current docs unfortunately)... I need to think about the problem again, and it might change this discussion...